nbclient/client.py

```diff
@@ -426,6 +483,7 @@ async def async_start_new_kernel_client(self) -> KernelClient:
             await self._async_cleanup_kernel()
             raise
         self.kc.allow_stdin = False
+        await run_hook(self.on_execution_start)
```

Comment:
```suggestion
        await run_hook(self.on_execution_start, notebook=self.nd)
```

What about something like this so that people have access to the NotebookNode before execution?